src/com/ms/silverking/cloud/zookeeper/ZooKeeperExtended.java

```diff
@@ -42,1024 +43,1081 @@
+      this.paths = paths;
+      this.dataMap = new ConcurrentHashMap<>();
+      this.errorKeys = new ConcurrentSkipListSet<>();
+      this.lock = new ReentrantLock();
+      this.cv = lock.newCondition();
+      this.watcher = watcher;
+      this.observer = observer;
+      retrySW = new SimpleStopwatch();
+      timeoutSW = new SimpleStopwatch();
+      displayMissingSW = new SimpleStopwatch();
     }
-    
-    public Map<String,byte[]> getByteArrays(Set<String> paths, Watcher watcher) throws KeeperException {
-        AsyncGet    aGet;
-        
-        aGet = new AsyncGet(paths, watcher, null);
-        aGet.issueGets();
-        aGet.waitForCompletion();
-        return aGet.getDataMap();
+
+    AsyncGet(Set<String> paths) {
+      this(paths, null, null);
     }
-    
-    public Map<String,byte[]> getByteArrays(String basePath, Set<String> children, Watcher watcher, CancelableObserver observer) throws KeeperException {
-        ImmutableSet.Builder<String>    childrenPaths;
-        AsyncGet    aGet;
-        
-        childrenPaths = ImmutableSet.builder();
-        for (String child : children) {
-            childrenPaths.add(basePath +"/"+ child);
-        }
-        
-        aGet = new AsyncGet(childrenPaths.build(), watcher, observer);
-        aGet.issueGets();
-        aGet.waitForCompletion();
-        
-        return stripBasePathFromKeys(aGet.getDataMap(), basePath);
+
+    Map<String, byte[]> getDataMap() {
+      return dataMap;
     }
-    
-    public Map<String,byte[]> getByteArrays(String basePath, Set<String> children) throws KeeperException {
-        return getByteArrays(basePath, children, null, null);
+
+    void issueGets() {
+      for (String path : paths) {
+        //System.out.println("g: "+ path);
+        getData(path, watcher, this, null);
+      }
+      retrySW.reset();
     }
-    
-    private Map<String, byte[]> stripBasePathFromKeys(Map<String, byte[]> rawMap, String basePath) {
-        ImmutableMap.Builder<String, byte[]>    strippedMap;
-        
-        strippedMap = ImmutableMap.builder();
-        for (Map.Entry<String, byte[]> entry : rawMap.entrySet()) {
-            strippedMap.put(stripBasePath(entry.getKey(), basePath), entry.getValue());
+
+    public void issueGetsForIncomplete() {
+      if (observer == null || observer.isActive()) {
+        for (String path : paths) {
+          if (!dataMap.containsKey(path) && !errorKeys.contains(path)) {
+            Log.warning("Issuing sync zk get: ", path);
+            try {
+              byte[] data;
+
+              data = getData(path, watcher, null);
+              if (data != null) {
+                Log.warning("Sync zk get success: ", path);
+                dataMap.put(path, data);
+              } else {
+                Log.warning("Sync zk get failed to retrieve data; adding to error keys: ", path);
+                errorKeys.add(path);
+              }
+            } catch (KeeperException | InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)